app/cdap/components/PipelineDiff/util/fetch.ts

```diff
@@ -0,0 +1,126 @@
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+import { Observable } from 'rxjs/Observable';
+import { MyReplicatorApi } from 'api/replicator';
+import { MyPipelineApi } from 'api/pipeline';
+import { AvailablePluginsMap, IPipelineConfig, IPipelineStage } from '../types';
+import {
+  createPluginInfo,
+  getAvailabePluginsMapKeyFromArtifact,
+  getPluginKeyFromPluginProps,
+} from '../util/helpers';
+
+type ApiCall<TOptions = any, TResponse = any> = (options: TOptions) => Observable<TResponse>;
+
+interface IFetchPipelineOptions {
+  namespace: any;
+  appId: any;
+  topVersion: any;
+  bottomVersion: any;
+}
+interface IFetchPipelineRepsonse {
+  topPipelineConfig: IPipelineConfig;
+  bottomPipelineConfig: IPipelineConfig;
+}
+/**
+ * @param options required options to fetch two pipeline versions
+ * @returns Observable that provides the two typed pipelines
+ */
+export const fetchPipelineConfig: ApiCall<IFetchPipelineOptions, IFetchPipelineRepsonse> = (
+  options
+) => {
+  const { namespace, appId, topVersion, bottomVersion } = options;
+  return new Observable<IFetchPipelineRepsonse>((subscriber) => {
+    Observable.forkJoin(
+      MyPipelineApi.getAppVersion({
+        namespace,
+        appId,
+        version: topVersion,
+      }),
+      MyPipelineApi.getAppVersion({
+        namespace,
+        appId,
+        version: bottomVersion,
+      })
+    ).subscribe((res: [any, any]) => {
+      const [topRes, bottomRes] = res;
+      const topPipelineConfig: IPipelineConfig = JSON.parse(topRes.configuration);
+      const bottomPipelineConfig: IPipelineConfig = JSON.parse(bottomRes.configuration);
+      subscriber.next({ topPipelineConfig, bottomPipelineConfig });
+    });
+  });
+};
+
+interface IFetchExtraPluginPropertiesOptions {
+  namespace: any;
+  stages: IPipelineStage[];
+}
+
+interface IFetchExtraPluginPropertiesResponse {
+  availablePluginsMap: AvailablePluginsMap;
+}
+/**
+ * Fetches extra plugin properties and generates an available plugins map, provieds the
+ * map through an observable.
+ * @param options required options to fetch the extra plugin properties
+ * @returns Observable that provides a available plugins map
+ */
+export const fetchExtraPluginProperties: ApiCall<
+  IFetchExtraPluginPropertiesOptions,
+  IFetchExtraPluginPropertiesResponse
+> = (options) => {
+  const { stages, namespace } = options;
+
+  const availablePluginsMap = {};
+  const reqBody = [];
+  stages.forEach((stage) => {
+    const plugin = stage.plugin;
+    const pluginInfo = createPluginInfo(plugin);
+
+    reqBody.push(pluginInfo.info);
+    availablePluginsMap[pluginInfo.key] = {
+      pluginInfo: stage.plugin,
+    };
+  });
+
+  return new Observable<IFetchExtraPluginPropertiesResponse>((subscriber) => {
+    MyReplicatorApi.batchGetPluginsWidgets({ namespace }, reqBody).subscribe((res) => {
```

Comment:
It doesn't look right using replicatorApi.. So Replicator is the subset of all CDAP pipelines. But it would be better to copy over this endpoint to `pipeline.js` instead of calling replicatorApi, since this is not replicator. Some duplication is ok.

Reply:
Done.